src/backend/validate.ts

```diff
@@ -0,0 +1,60 @@
+import { NegatableOption, RequestSpec } from '../RequestSpec';
+import { Edition } from '../types';
+import { labelForRequestSpecKey } from '../ui/utils';
+
+// Type predicate
+// eslint-disable-next-line
+const isNegatableOption = (value: any): value is NegatableOption => {
+  return typeof value.negated === 'boolean';
+};
+
+const isDefaultValue = (value: string | NegatableOption | string[] | undefined): boolean => {
+  if (typeof value === 'undefined') {
+    return true;
+  }
+  if (typeof value === 'string') {
+    return value === '';
+  }
+  if (isNegatableOption(value)) {
+    return value.value !== undefined || value.value !== '';
+  }
+  if (Array.isArray(value)) {
+    return value.length > 0;
+  } else {
+    throw new Error(`Impossible, Should never happen: isDefaultValue got an Argument of type ${typeof value}`);
+  }
+};
+
+type PotentiallyRequiredKeys = 'graph_type' | 'aggregation' | 'site' | 'service' | 'host_name' | 'graph';
+
+export const getRequiredFields = (edition: Edition) => {
+  const result: PotentiallyRequiredKeys[] = ['graph_type', 'aggregation'];
+  if (edition === 'CEE') {
+    result.push('graph');
+  } else {
+    result.push('site', 'service', 'host_name', 'graph');
+  }
+  return result.sort();
+};
+
+const missingRequiredFields = (rq: Partial<RequestSpec>, edition: Edition): string[] => {
```

Comment:
have you thought about hard coding it for each key? would be not elegant code, but easy to read and understand and we would not have to guess the type (because we already know it by the name of the key?)?

another idea: introduce a function `getDefaultValue(key)` that is also called when adding a filter, and simply compare elements in the request spec with this default value?

Reply:
I like the first approach a lot, especially in combination with the comment above. Was thinking way more complicated that I needed here.